app/controllers/encrypt_controller.rb

```diff
@@ -179,6 +178,44 @@ def stats
     render json: success_json.merge(encrypted_pms_count: [pms_count, keys_count].max)
   end
 
+  # Get all topic keys for current user.
+  #
+  # Returns all keys to all encrypted PMs.
+  def show_all_keys
+    encrypted_topic_users = EncryptedTopicsUser.where(user: current_user)
+    topic_keys = encrypted_topic_users.map { |etu| [etu.topic_id, etu.key] }.to_h
+    render json: success_json.merge(topic_keys: topic_keys)
+  end
+
+  # Updates all keys for current user.
+  #
+  # Params:
+  # +keys+::    All topic keys
+  # +public+::  Serialized public identity
+  #
+  # Returns status code 200 if all keys are updated or status code 400 if not
+  # all keys have a replacement.
+  def update_all_keys
+    ActiveRecord::Base.transaction do
+      raise Discourse::InvalidParameters.new(:public) if params[:public].blank?
+
+      a = EncryptedTopicsUser.where(user: current_user).map { |tk| tk.topic_id }
+      b = (params[:keys] || {}).keys.map(&:to_i)
+      raise Discourse::InvalidParameters.new(:keys) if a.sort != b.sort
+
+      user_key = UserEncryptionKey.find_or_initialize_by(user_id: current_user.id)
+      user_key.encrypt_public = params[:public]
+      user_key.encrypt_private = nil
+      user_key.save!
+
+      params[:keys].each do |topic_id, key|
+        EncryptedTopicsUser.find_by(user: current_user, topic_id: topic_id).update!(key: key)
```

Comment:
You don't need to `find` and then `update`, you can do it in one SQL query with something like this

```ruby
EncryptedTopicsUser.where(user: current_user, topic_id: topic_id).update_all(key: key)
```